Minor fixes in design and readme (#532)
oam-dev/kubevela: docs/design.md

- Heroku-like (_in terms of both user experience and functionality_). 
  - and I prefer to build my own version with OSS tools, particularly, with Kubernetes (for obvious reason).
- Easy to build. 

>   - I am too busy to reinvent any wheel, I want to reuse every capability in Kubernetes community as possible,  with minimal effort. Writing some simple CRD and controllers is fine, but please, just the simple ones.

Reviewer: reuse every capability -> reuse as many capabilities

Author: Done, all fixed!

>   - I am too busy to reinvent any wheel, I want to reuse as many capabilities in Kubernetes community as possible, with minimal effort. Writing some simple CRD and controllers is fine, but please, just the simple ones.

Reviewer 2: ```suggestion
  - I am too busy to reinvent any wheel, I want to reuse the existing capabilities in Kubernetes community with minimal effort. Writing some simple CRD and controllers is fine, but please, just the simple ones.
```

- Powerful and highly extensible.
  - I don't want to lock users with restricted abstractions and capabilities like traditional PaaS or FaaS/Serverless. I love Kubernetes and what it has enabled. So in terms of capability, I hope my platform is fully open and has unlimited possibilities like Kubernetes itself, rather than another opinionated close system like traditional PaaS.

## Core Principles

> For developers, KubeVela out-of-the-box provides a PaaS-like experience that enables better productivity (i.e. focus, self-service, on-demand, and fast-feedback).

Reviewer 2: ```suggestion
For developers, KubeVela provides a out-of-the-box PaaS-ish experience that enables better productivity (i.e. focus, self-service, on-demand, and fast-feedback).
```

> For platform builders, KubeVela works like a special Kubernetes "distro" or extensible PaaS core that could be used to build something more complex on top of. It enables platform builders to integrate any existing capabilities in Kubernetes ecosystem with minimal effort, or develop a new capability at ease in a standardized and Kubernetes-native approach.

Reviewer 2: ```suggestion
For platform builders, KubeVela works like a special Kubernetes "distro" or extensible PaaS core that could be used to build something more complex on top of. It enables platform builders to integrate existing capabilities in Kubernetes ecosystem with minimal effort, and/or develop a new capability at ease in a standard, Kubernetes-native approach.
```

## Design Details

### 1. Application Centric

Keywords: _focus_, _self-service_.

> KubeVela intends to ensure developers only think in terms of _application_, not container or infrastructure. Lacking application context impacts the developer experience and significantly raised the bar to adopt cloud native technology. We believe _application_ is the natural mindset of developers and it's the core concept KubeVela should expose.

Reviewer 2: ```suggestion
KubeVela intends to ensure developers only think in the scope of _application_, not container or infrastructure. Lacking application context impacts the developer experience and raises the bar to adopt cloud native technology. We believe _application_ is the natural mindset for developers and it's the core concept KubeVela should expose.
```

![alt](../resources/app-centric.png)

> In KubeVela, developers will only push code, deploy application in developer facing primitives, and claim operations policies as part of the application definition. Thus, KubeVela choose to:

Reviewer 2: ```suggestion
In KubeVela, it is only push code, deploy application in developer facing primitives, and claim operations policies that developers would do. Thus, KubeVela choose to:
```

1. Introduce _application_ as its first class API.
2. Build the whole system around "application", i.e. model capabilities of Kubernetes as application configuration, with clarity and manageability.

#### Solution

One common solution to introduce such concept of "application" to Kubernetes is creating an "application CRD". However, as an extensible PaaS core, KubeVela intends to support any application type and any operation feature in Kubernetes ecosystem, thus instead of creating a in-house and "monolithic" CRD, KubeVela adopts [Open Application Model (OAM)](https://github.com/oam-dev/spec) as its application definition. This is empowers KubeVela with high extensibility since every workload and operation feature in OAM is modeled by independent definition object with full freedom to extend. Besides, the model in OAM is also well-defined as it defines micro-services application by default, and models operation features as part of the application (i.e. `Traits`).

### 2. Capability Oriented Architecture

Keywords: _on-demand_.

> KubeVela intends to make its every capability a standalone "plug-in" in terms of implementation and configuration. This enables the extensibility of the platform, and gives platform builders full freedom to extend KubeVela into anything else they want (e.g. database PaaS, AI PaaS, etc.).

Reviewer 2: ```suggestion
Every capability in KubeVela is designed as a standalone "plug-in" in terms of implementation and configuration. This enables the extensibility of the platform, and gives platform builders full freedom to extend KubeVela into anything else they want (e.g. database PaaS, AI PaaS, etc.).
```
